payjoin/src/receive/mod.rs

```diff
@@ -437,45 +530,74 @@ impl ProvisionalProposal {
             .map(|input| input.sequence)
             .unwrap_or_default();
 
-        // Add the value of new receiver input to receiver output
-        let txo_value = txo.value;
-        let vout_to_augment =
-            self.owned_vouts.choose(&mut rand::thread_rng()).expect("owned_vouts is empty");
-        self.payjoin_psbt.unsigned_tx.output[*vout_to_augment].value += txo_value;
-
-        // Insert contribution at random index for privacy
+        // Insert contributions at random indices for privacy
         let mut rng = rand::thread_rng();
-        let index = rng.gen_range(0..=self.payjoin_psbt.unsigned_tx.input.len());
-        self.payjoin_psbt
-            .inputs
-            .insert(index, bitcoin::psbt::Input { witness_utxo: Some(txo), ..Default::default() });
-        self.payjoin_psbt.unsigned_tx.input.insert(
-            index,
-            bitcoin::TxIn {
-                previous_output: outpoint,
-                sequence: original_sequence,
-                ..Default::default()
-            },
-        );
-    }
+        let mut inputs_amount = Amount::ZERO;
+        for (outpoint, txo) in inputs.into_iter() {
+            inputs_amount += txo.value;
+            let index = rng.gen_range(0..=self.payjoin_psbt.unsigned_tx.input.len());
+            payjoin_psbt.inputs.insert(
+                index,
+                bitcoin::psbt::Input { witness_utxo: Some(txo), ..Default::default() },
+            );
+            payjoin_psbt.unsigned_tx.input.insert(
+                index,
+                bitcoin::TxIn {
+                    previous_output: outpoint,
+                    sequence: original_sequence,
+                    ..Default::default()
+                },
+            );
+        }
 
-    pub fn is_output_substitution_disabled(&self) -> bool {
-        self.params.disable_output_substitution
-    }
+        // Add the receiver change amount to the receiver change output, if applicable
+        let min_input_amount = self.min_input_amount();
+        if inputs_amount >= min_input_amount {
+            let change_amount = inputs_amount - min_input_amount;
+            let vout_to_augment =
+                self.owned_vouts.choose(&mut rand::thread_rng()).expect("owned_vouts is empty");
+            payjoin_psbt.unsigned_tx.output[*vout_to_augment].value += change_amount;
```

Comment:
Note that this prevents the receiver from losing money, but it also prevents them from calculating their own miner fees and subtracting them from the provided outputs value, since any deficit would be added back here.

I noticed that while trying to manually apply fees for the receiver in the UTXO consolidation test.

Reply:
"manually" meaning at output substitution/input contribution and before `apply_fees` is called?

I suppose the decision to be made is whether or not that is a desirable property of our receiver construction api

Reply:
Right, if you subtract fees from a receiver output preemptively in the output substitution step it gets overriden here.

I think that's probably fine and fee contribution should most likely be an explicit step later on (`apply_fees`). My only concern with that approach is I think the API should minimize "surprises", so it should be very clear that the amounts provided in the output substitution step are not final and may increase or decrease based on excess input amount and required fees.